### PR TITLE
Try to add a MacOS test runner

### PR DESCRIPTION
**Pull request by ericniebler**

### CODE REVIEW

File: .github/workflows/ci.cpu.yml [hunk 1]
```diff
@@ -114,3 +114,41 @@ jobs:
...
+        run: |
+          brew update
+          brew install ninja
+        shell: bash
+
+      - name: Build and test CPU schedulers (MacOS)
+        shell: bash
+        run: |
+          mkdir build
+          cmake -S. -Bbuild -GNinja -DCMAKE_BUILD_TYPE=${{ matrix.build }} -DCMAKE_CXX_COMPILER=${{ matrix.compiler }} -DCMAKE_CXX_STANDARD=20
```
Comment (ericniebler) [inline .github/workflows/ci.cpu.yml:143]:
```suggestion
          cmake -S. -Bbuild -GNinja -DCMAKE_BUILD_TYPE=${{ matrix.build }} -DCMAKE_CXX_COMPILER=${{ matrix.compiler }} -DCMAKE_CXX_FLAGS=-stdlib=libc++ -DCMAKE_CXX_STANDARD=20
```

File: .github/workflows/ci.cpu.yml [hunk 2]
```diff
@@ -114,3 +114,41 @@ jobs:
...
+        run: |
+          brew update
+          brew install ninja
+        shell: bash
+
+      - name: Build and test CPU schedulers (MacOS)
+        shell: bash
+        run: |
+          mkdir build
+          cmake -S. -Bbuild -GNinja -DCMAKE_BUILD_TYPE=${{ matrix.build }} -DCMAKE_CXX_COMPILER=${{ matrix.compiler }} -DCMAKE_CXX_FLAGS=-stdlib=libc++ -DCMAKE_CXX_STANDARD=20
```
Comment (ericniebler) [inline .github/workflows/ci.cpu.yml:143]:
```suggestion
          cmake -S. -Bbuild -GNinja \
            -DCMAKE_BUILD_TYPE=${{ matrix.build }} \
            -DCMAKE_CXX_COMPILER=${{ matrix.compiler }} \
            -DCMAKE_CXX_FLAGS=-stdlib=libc++ \
            -DCMAKE_CXX_STANDARD=20 \
            -DCMAKE_EXE_LINKER_FLAGS="-stdlib=libc++ -lc++abi"
```

File: .github/workflows/ci.cpu.yml [hunk 3]
```diff
@@ -114,3 +114,46 @@ jobs:
...
+      - name: Install dependencies
+        run: |
+          brew update
+          brew install ninja
+        shell: bash
+
+      - name: Build and test CPU schedulers (MacOS)
+        shell: bash
+        run: |
+          mkdir build
```
Comment (ericniebler) [inline .github/workflows/ci.cpu.yml:142]:
```suggestion
          set -eo pipefail
          mkdir build
```

File: .github/workflows/ci.cpu.yml [hunk 4]
```diff
@@ -114,3 +114,46 @@ jobs:
...
+          cmake -S. -Bbuild -GNinja \
+            -DCMAKE_BUILD_TYPE=${{ matrix.build }} \
+            -DCMAKE_CXX_COMPILER=${{ matrix.compiler }} \
+            -DCMAKE_CXX_FLAGS=-stdlib=libc++ \
+            -DCMAKE_CXX_STANDARD=20 \
+            -DCMAKE_EXE_LINKER_FLAGS="-stdlib=libc++ -lc++abi"
```
Comment (ericniebler) [inline .github/workflows/ci.cpu.yml:148]:
```suggestion
            -DCMAKE_EXE_LINKER_FLAGS="-stdlib=libc++ -lc++ -lc++abi"
```

**Comment by ericniebler** [comment]
@morinmorin thank you very much for this analysis. it is very helpful. that `__state` object should not be moved at all. the fact that it is getting copied leads me to think this is a problem with apple clang's guaranteed copy elision.

**Comment by morinmorin** [comment]
> debug is now passing all tests.

Oh, this is so nice!

It looks like I ran tests in release mode (I didn't notice that building stdexec with CMake defaults to release mode). I confirmed this fix also works on my system as on GitHub Action macOS runners.

**Comment by morinmorin** [comment]
If we add `optnone` attribute to `get_id()` function in the `test_task.cpp` (see [here](https://clang.llvm.org/docs/AttributeReference.html#id598)), i.e.
```
[[clang::optnone]] int get_id() {
  return __thread_id;
}
```
all the tests pass in release/debug modes.

**Comment by ericniebler** [comment]
Wow, thanks!

**Pull request merged**